### ISSUE TITLE
Merge api and sentinel into webapp in git

### ISSUE THREAD

**Issue by garethbowen** [medic/medic-webapp#3359]
As discussed in [this doc](https://docs.google.com/document/d/1bUfGjviRxWRsbi_z0BXPbmqBfSXzLhPGgUsRZw6hpPs/edit) we've decided to merge the api and sentinel repos into webapp so long as the git histories can be retained.

Do the repo merge and where desirable merge or raise further issues to merge tests, dependencies, gruntfiles, and anything else you can think of.

**Comment by garethbowen** [comment]
Node modules that are required at runtime will have to be bundled up for publishing so will have to somehow end up in the api or sentinel directory respectively. It's probably best to leave these dependencies unmerged, but I think most if not all of the dev dependencies could be merged into a single root level package.

**Comment by mandric** [comment]
@garethbowen and @SCdF I know we talked about this but I was still trying to think about the downsides of using npm for this.  Is basically the only downside to keeping things separate and treating sentinel and api like any other npm module is that it makes it hard to do branching/development because you have to potentially branch multiple projects?  This was the main takeaway I can remember from our conversation.

**Comment by SCdF** [comment]
@mandric take a look at the document: https://docs.google.com/document/d/1bUfGjviRxWRsbi_z0BXPbmqBfSXzLhPGgUsRZw6hpPs/edit then apply each of the items in the problems we are trying to solve against the theoretical npm solution. We did it on the call, and I don't want to type it all out again, but basically it didn't fix those problems, and it actually made releasing a bit more complicated because it adds two more changes you need to make to lock down versions, as well as having to build and release api and sentinel before you build and release medic-webapp, which introduces an ordering problem we didn't have before.

To be clear, we should still split out the "api" parts of api into a separate module / repo / npm package / whatever, that is a good idea. It's separate to this discussion though.

**Comment by estellecomment** [comment]
When we do that, we could move the api e2e tests into the webapp e2e tests. They both need couch + api running, it ends up being the same setup. 
The migration tests might be different though (in `integration` dir)

**Comment by mandric** [comment]
Well here's my thoughts...  If anything I think medic-api and medic-sentinel should be merged.  Rather than listening to  changes feed we can make changes to records as data comes in.  I'm requesting a little time to do some refactor first before we do this because it's a big change.

> There are at least 8 interactions you need to get right and consistent to release a new version

This can be automated if needed.  Even with one repo we still have lots of interaction.

> When commit against api or sentinel you have to remember to add medic/medic-webapp before the ticket number so it links right

That's a bad expectation, always use the full urls.

> We only run integration tests against whatever is statically linked in the submodules, so api and sentinel master is often not being exercised

Link to GitHub master branch in package.json, so any commit to webapp master also tests api/sentinel master branches.

> Changing api or sentinel doesn’t trigger webapp to run integration tests

I'm not sure this is a practical problem.  Integration tests will run on the next commit which typically happens every day if not multiple times a day.  If you make a change in one of webapp's deps and you really want to run tests against webapp again then restart the travis build manually.  

Every module should have its own coverage, if there is an e2e test that fails in webapp because of a bug in a dep then it's  also a bug in the dep if there is not a test that is failing/covers that. 

> To run all tests locally you need to run three grunt tasks from three different directories

I don't understand why that's a problem, that's what Grunt is for to automate these types of tasks, like cd into a subdirectory and run something.  Seems like standard practice.

> And again, if you have checked out api and sentinel in places other than the submodules, the integration tests won’t be run over your code

Use `npm link` if you are developing on a dep of webapp, then npm will use your local copy in webapp.

**Comment by garethbowen** [comment]
> That's a bad expectation, always use the full urls.

That's what I've been trying to do recently. Firstly it's easy to forget to do this. Secondly it's really easy to typo `medci/mdeic-webpap#xxxx`. Thirdly it's just unnecessary boilerplate if we make the proposed change.

> This can be automated if needed. Even with one repo we still have lots of interaction.

Not entirely. It can certainly be improved through automation but it would be eliminated entirely by merging, _and_ we wouldn't need to support custom automation code. Elimination trumps automation IMO.

> Link to GitHub master branch in package.json, so any commit to webapp master also tests api/sentinel master branches.

Having GH dependencies in packages.json is really bad practice which causes performance and stability issues because NPM needs to check the repo every time you install. Also IIRC it prohibits us from using tools like yarn for better dependency management (npm shrinkwrap sucks so bad).

> I'm not sure this is a practical problem. Integration tests will run on the next commit which typically happens every day if not multiple times a day. 

This is true, however we want these integration tests to fail on the feature branch before being merged into master. If I make a change to sentinel in a feature branch that would break an e2e test in webapp I won't find out until it's merged into master. Furthermore it won't fail a build until someone else commits to master so it won't be obvious which change broke the build.

> If you make a change in one of webapp's deps and you really want to run tests against webapp again then restart the travis build manually.

I'm sure nobody merges code to master they know is broken - the whole point of CI is to run _every time_ (ie: continuously) to pick up something you've missed.

> Every module should have its own coverage, if there is an e2e test that fails in webapp because of a bug in a dep then it's also a bug in the dep if there is not a test that is failing/covers that.

I don't agree. The point of integration tests (e2e in your example) is you're testing how two modules work together. If an integration test fails it's not necessarily because there's a bug or lack of coverage in the dependency, it may indicate an issue with how the parent is using the dependency.

> I don't understand why that's a problem, that's what Grunt is for to automate these types of tasks, like cd into a subdirectory and run something.

Yes we could do something like this. It would require everyone has the same dir structure but it can be done.

> Use npm link if you are developing on a dep of webapp, then npm will use your local copy in webapp.

We're not talking about npm dependencies. Publishing api and sentinel to npm is a whole other issue.

@mandric On the call where you, @SCdF and I discussed this we all agreed (IIRC) that the proposed solution was better than the existing submodule system and that we should give it a try. We can go back to the submodule system or switch to publishing to npm at a later date if we want to. The points you raised above were discussed in that call and it's useful to have them documented (thanks!) but I don't see anything which changes my mind. In short, the proposed change will make my development experience significantly easier and more reliable through fewer bugs and broken builds. Are you willing to disagree and commit on this?

As always, I'm happy to talk here, on Slack, in Hangouts, or in person ( :beer: ).

**Comment by mandric** [comment]
> > That's a bad expectation, always use the full urls.
> That's what I've been trying to do recently. Firstly it's easy to forget to do this. Secondly it's really easy to typo medci/mdeic-webpap#xxxx. Thirdly it's just unnecessary boilerplate if we make the proposed change.

If you don't want to do repetitive things in your git commit messages then configure git to help nudge us in the right direction for commit formats: https://git-scm.com/book/en/v2/Customizing-Git-Git-Configuration


> > > There are at least 8 interactions you need to get right and consistent to release a new version

> > This can be automated if needed. Even with one repo we still have lots of interaction.

> Not entirely. It can certainly be improved through automation but it would be eliminated entirely by merging, and we wouldn't need to support custom automation code. Elimination trumps automation IMO.

I don't believe merging eliminates this.

> > Link to GitHub master branch in package.json, so any commit to webapp master also tests api/sentinel master branches.

> Having GH dependencies in packages.json is really bad practice which causes performance and stability issues because NPM needs to check the repo every time you install. Also IIRC it prohibits us from using tools like yarn for better dependency management (npm shrinkwrap sucks so bad).

We already have GH dependencies in our build system (see openrosa-formlist dep) and it hasn't been an issue.  Github is about as reliable as npm in my experience.  So there is not much difference there.  


> > I'm not sure this is a practical problem. Integration tests will run on the next commit which typically happens every day if not multiple times a day.


> This is true, however we want these integration tests to fail on the feature branch before being merged into master. If I make a change to sentinel in a feature branch that would break an e2e test in webapp I won't find out until it's merged into master. Furthermore it won't fail a build until someone else commits to master so it won't be obvious which change broke the build.

> > If you make a change in one of webapp's deps and you really want to run tests against webapp again then restart the travis build manually.

> I'm sure nobody merges code to master they know is broken - the whole point of CI is to run every time (ie: continuously) to pick up something you've missed.

> > Every module should have its own coverage, if there is an e2e test that fails in webapp because of a bug in a dep then it's also a bug in the dep if there is not a test that is failing/covers that.

> I don't agree. The point of integration tests (e2e in your example) is you're testing how two modules work together. If an integration test fails it's not necessarily because there's a bug or lack of coverage in the dependency, it may indicate an issue with how the parent is using the dependency.

> > I don't understand why that's a problem, that's what Grunt is for to automate these types of tasks, like cd into a subdirectory and run something.

> Yes we could do something like this. It would require everyone has the same dir structure but it can be done.

Bascially if you want to test a module against a project that depends on that module, then create a branch and a PR against that project, just like any other change.  Yes it's a little more work but a small price for maintaining separation.

Having one huge repo might developer experience but my fear is that it leads to code that is messier.  It promotes interdependency between parts of the code base that should be separate.  It promotes monolithic architectures that ultimately leads to software that is difficult for third parties to understand and develop on.  I am generally in favor of breaking out and creating more separate modules/repos that are easier to maintain and integrate.  It is the responsibility of the project that uses various dependencies to solve the problem of integration testing.  Merging modules into one repo is not a good solution for that.  We should work on a solution for integration testing that also preserves module independence.

I am still hearing that the main problem here is working on a feature branch of webapp that also touches sentinel and/or api. I don't think the solution for that involves merging repos and we should try to find a proper solution for that.  I personally don't have a problem with maintaining two feature branches in two repos.  Merging repos to alleviate this minor bump in a developer's workflow seems absurd?

**Comment by garethbowen** [comment]
@mandric I don't think you understand exactly how much of a pain this is. It's certainly not a "minor bump". In the last week we've had problems with submodule revisions disappearing due to depending on a branch which is then merged and deleted, and problems with releasing 2.12.0, and features requiring 3 PRs that are all co-dependent so have to be merged at the same time. And that's just in the last week.

Given that you're unwilling to give the proposed solution a try I'm going to close this. There's simply no point going over the same old arguments again.

**Comment by mandric** [comment]
I sense your frustration.  I swear I'm just trying to understand the pain and express my point of view.  This is a significant change and it's quite a bit of work to address all points of impact.  I was hoping we could use our base process of PRs to help us alleviate some pain, because I don't think we've really been doing that with submodule commits.  My theory is if we treat submodule changes like any other change it will alleviate much of the pain we've experienced thus far.  If the team sees no way forward except merging repos (which is also quite a bit of work) I'm fine with conceding.

**Comment by garethbowen** [comment]
This is the process I followed...

```
# Submodule merging

## Move all submodule files into a folder

cd ../<submodule_dir>
git checkout -b "3359-merge-submodules-into-webapp"
mkdir <submodule_name>
git mv <all_files> <submodule_name>
git commit -m "Merge submodules into webapp." -m "medic/medic-webapp#3359"
git push --set-upstream origin 3359-merge-submodules-into-webapp

## Deinit the submodule

cd ../medic-webapp
git submodule deinit -f -- <submodule_name>
rm -rf .git/modules/<submodule_name>
git rm -f <submodule_name>

## Merge submodule into webapp

cd ../medic-webapp
git remote add -f old_<submodule_name> <submodule_repo_url>
git merge old_<submodule_name>/3359-merge-submodules-into-webapp
git push --set-upstream origin 3359-merge-submodules-into-webapp
```

**Comment by garethbowen** [comment]
I merged this last night but immediately noticed I couldn't view the file histories in GH. The log in the CLI was complete so it looks like it's just a bug in the GH UI. FWIW the GH blame UI showed all the changes as expected. Stefan and I then force reverted the merge from master.

I've lodged a support request with GH to see if there's some way to show the history. Failing that I'll check with the other developers to see if this is a blocker or not.
